### DISCUSSION TITLE
Converting a (default) str to an HttpUrl

### DISCUSSION THREAD

**Discussion by jwodder** [pydantic/pydantic#2492] [Question] [answered]
Consider the following code:

```python
from pydantic import BaseModel, HttpUrl

class MyModel(BaseModel):
    url: HttpUrl = "https://www.example.com"

m = MyModel()
print(repr(m.url))
```

This prints out `'https://www.example.com'`, indicating that the default value is not converted to an `HttpUrl`, which is undesirable if the rest of your code depends on being able to access various properties of `m.url`.

I figured I could work around this by explicitly casting the default to an `HttpUrl`, like so:

```python
from pydantic import BaseModel, HttpUrl

class MyModel(BaseModel):
    url: HttpUrl = HttpUrl("https://www.example.com")

m = MyModel()
print(repr(m.url))
```

However, this fails with `TypeError: __init__() needs keyword-only argument scheme`.  As far as I can tell, in order to instantiate an `HttpUrl` directly, one has to explicitly provide all of the nonempty fields; thus:

```python
from pydantic import BaseModel, HttpUrl

class MyModel(BaseModel):
    url: HttpUrl = HttpUrl(
        "https://www.example.com",
        scheme="https",
        host="www.example.com",
        tld="com",
        host_type="domain",
    )

m = MyModel()
print(repr(m.url))
```

This works, but it seems excessive.  Is there no easier way to convert a (default value) string to an `HttpUrl`?

**Comment by jwodder** [comment, marked as answer]
I've found that `url: HttpUrl = parse_obj_as(HttpUrl, "https://www.example.com")` works, but I still wonder if that's the intended way to create an `HttpUrl`.

**Comment by PrettyWood** [comment]
Hi @jwodder 
You can probably set `validate_all = True` in the `Config` to force validation (hence coercion) of default values

**Comment by jwodder** [comment]
@PrettyWood That works as well.

**Comment by PrettyWood** [comment]
@jwodder Is your issue solved or were you expecting something else? if it is solved please mark this discussion as answered 👍
